### PR TITLE
Complex patterns

### PR DESCRIPTION
**Pull request by angrykoala**
# Description
Implements complex patterns in cypher builder following the RFC #2483 

Updates graphql library translation to support the new syntax

## Complexity

Complexity: Medium

### CODE REVIEW

**Comment by neo4j-team-graphql** [comment]
# Performance Report

No Performance Changes


<details>
    <summary>Show Full Table</summary>
    
| name | dbHits | old dbHits | time (ms) | old time (ms) | maxRows |
| ---- | ------ | ---------- | --------- | ------------- | ------- |
|  aggregations.TopLevelAggregate | 3403 | 3403 | 34 | 48 | 1134 |
|  aggregations.NestedAggregation | 13514 | 13514 | 97 | 139 | 2174 |
|  aggregations.AggregationWithWhere | 10942 | 10942 | 69 | 102 | 2174 |
|  aggregations.NestedCountFromMovieToActors | 8694 | 8694 | 65 | 96 | 2174 |
|  aggregations.NestedCountFromActorsToMovie | 8900 | 8900 | 74 | 95 | 2174 |
|  aggregations.DeeplyNestedCount | 10054408 | 10054408 | 4780 | 5116 | 2008534 |
|  batch-create.BatchCreate | 3600 | 3600 | 95 | 125 | 600 |
|  connect.createAndConnect | 12419 | 12419 | 241 | 255 | 3003 |
|  connections.Connection | 13042 | 13042 | 221 | 134 | 2174 |
|  connections.NestedConnection | 33883 | 33883 | 121 | 169 | 4516 |
|  connections.ConnectionWithSort | 2277 | 2277 | 59 | 71 | 1040 |
|  connections.ConnectionWithSortAndCypher | 2277 | 2277 | 211 | 321 | 1040 |
|  create.SimpleMutation | 6 | 6 | 36 | 46 | 1 |
|  cypher-directive.TopLevelCypherDirective | 99525 | 99525 | 668 | 826 | 10168 |
|  cypher-directive.TopLevelCypherDirectiveWithColumnName | 127994 | 127994 | 225 | 314 | 10168 |
|  delete.SimpleDelete | 18361 | 18361 | 214 | 234 | 1040 |
|  delete.NestedDeleteInUpdate | 16779 | 16779 | 178 | 175 | 2040 |
|  query.SimpleQuery | 14120 | 14120 | 167 | 134 | 2174 |
|  query.QueryWhere | 8665 | 8658 | 36 | 55 | 2163 |
|  query.SimpleQueryWithNestedWhere | 8836 | 8829 | 48 | 56 | 2163 |
|  query.Nested | 10084290 | 10084290 | 9903 | 9990 | 2008534 |
|  query.NestedWithFilter | 37172 | 37172 | 143 | 153 | 4511 |
|  query.OrFilterOnRelationships | 36123 | 36035 | 164 | 192 | 1667 |
|  query.OrFilterOnRelationshipsAndNested | 30386 | 29724 | 189 | 252 | 1667 |
|  query.QueryWithNestedIn | 12300 | 12446 | 59 | 108 | 1201 |
|  query.NestedConnectionWhere | 8794 | 8794 | 54 | 75 | 2174 |
|  query.DeeplyNestedConnectionWhere | 8869 | 8898 | 79 | 108 | 2174 |
|  query.DeeplyNestedWithRelationshipFilters | 6251 | 6251 | 129 | 148 | 1134 |
|  query.DeeplyNestedWithRelationshipSingleFilters | 15438 | 15438 | 99 | 141 | 2174 |
|  query.Fulltext | 96 | 96 | 34 | 41 | 16 |
|  query.FulltextWithNestedQuery | 571 | 571 | 64 | 57 | 84 |
|  sorting.SortMultipleTypes | 2493 | 2493 | 134 | 85 | 1040 |
|  sorting.SortMultipleTypesWithCypherWithCypher | 1518 | 1406 | 259 | 325 | 1040 |
|  sorting.SortOnNestedFields | 13042 | 13042 | 91 | 79 | 2174 |
|  sorting.SortDeeplyNestedFields | 38965 | 38965 | 150 | 188 | 4516 |
|  sorting.SortWithTopLevelCypher | 2119 | 2119 | 138 | 223 | 1040 |
|  unions.SimpleUnionQuery | 321 | 321 | 65 | 62 | 35 |
|  unions.SimpleUnionQueryWithMissingFields | 293 | 293 | 57 | 69 | 35 |
|  unions.NestedUnion | 398553 | 398553 | 285 | 427 | 33033 |
|  unions.NestedUnionWithMissingFields | 372527 | 372527 | 310 | 306 | 33033 |
|  update.NestedUpdate | 2119 | 2119 | 40 | 53 | 1040 |

</details>
        
Old Schema Generation: 33.332s
Schema Generation: 33.228s

**Comment by neo4j-team-graphql** [comment]
Made some changes to the GraphQL Toolbox? Or made some changes to the core library perhaps?

Check out your changes at https://neo4j-graphql-toolbox-stage-2855.surge.sh

Use for instance a local dbms hosted on [Neo4j Desktop](https://neo4j.com/download/) or a cloud-hosted [Aura Free](https://neo4j.com/cloud/aura-free/) dbms.

File: packages/cypher-builder/src/pattern/PartialPattern.ts [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+        this.withType = false;
+        return this;
+    }
+
+    public withoutVariable(): this {
+        this.withVariable = false;
+        return this;
+    }
+
+    public withDirection(direction: "left" | "right" | "undirected"): this {
```
Comment (MacondoExpress) [inline packages/cypher-builder/src/pattern/PartialPattern.ts:65]: Potentially we could use an enum here

File: packages/cypher-builder/src/apoc/functions/RunFirstColumn.test.ts [hunk 2]
```diff
@@ -47,11 +47,7 @@ describe("RunFirstColumn", () => {
 
         const topQuery = new Cypher.Match(node).where(Cypher.eq(node.property("title"), param1));
 
-        const nestedPattern = node.pattern({
-            source: {
-                labels: false,
-            },
-        });
+        const nestedPattern = new Cypher.Pattern(node).withoutLabels();
```
Comment (MacondoExpress) [inline packages/cypher-builder/src/apoc/functions/RunFirstColumn.test.ts:50]: Way nicer!

**Review by MacondoExpress — APPROVED** [review]
Very impressive work here! Looks great to me :)

**Pull request merged**